Logs: Scope improvements

CodeBlanch opened the pull request:
## Issue 1: Scopes + Batching

There is an issue in the current scope API where if you capture a `LogRecord`, say for batching, the active scopes may have changed by the time the exporter runs.

* `LogRecord.BufferLogScopes` has been introduced to copy the active scopes into a buffer which can be inspected outside of the log lifecycle. This is now used by the `BatchLogRecordExportProcessor`. Made opt-in because a processor like `SimpleLogRecordExportProcessor` doesn't need the buffering and can use the `ForEachScope` call directly in the allocation-free form provided by the .NET API. **`LogRecord.ScopeProvider` is now cleared after `Processor.OnEnd` is called to prevent accidental misuse.**

## Issue 2: Working with Scopes

* The API today exposes the ability to loop over the scopes as `object`s. To be slightly more helpful we also now expose an enumerator which will return the individual `KeyValuePair<string, object>`s inside each scope. This makes it more aligned with the state parsing mechanism already in there. Can be done without allocation.

## Usage Example

### Sub-item Enumeration:

```csharp
                logRecord.ForEachScope(ProcessScopeItem, this);

                static void ProcessScopeItem(LogRecordScope scope, ConsoleLogRecordExporter exporter)
                {
                    foreach (KeyValuePair<string, object> scopeItem in scope)
                    {
                        exporter.WriteLine($"[Scope]:{scopeItem.Key.PadRight(RightPaddingLength)}{scopeItem.Value}");
                    }
                }
```

## API Changes

```csharp
// Existing
public sealed class LogRecord
{
   // Signature changed.
   public void ForEachScope<TState>(Action<LogRecordScope, TState> callback, TState state);
}

// New
public readonly struct LogRecordScope
{
        public object Scope { get; }

        public Enumerator GetEnumerator() => new Enumerator(this.Scope);

        // For enumerating sub-items on the scope.
        public struct Enumerator : IEnumerator<KeyValuePair<string, object>>
}
```

## TODOs

* [X] Unit tests.
* [X] `CHANGELOG.md` updated for non-trivial changes
* [ ] Changes in public API reviewed

CodeBlanch commented on src/OpenTelemetry/OpenTelemetry.csproj:
```diff
@@ -7,11 +7,13 @@
   <ItemGroup>
     <PackageReference Include="System.Collections.Immutable" Version="$(SystemCollectionsImmutablePkgVer)" />
     <PackageReference Include="System.Reflection.Emit.Lightweight" Version="$(SystemReflectionEmitLightweightPkgVer)" />
+    <PackageReference Include="System.Runtime.CompilerServices.Unsafe" Version="5.0.0" />
```
Note: This is already referenced by [System.Diagnostics.DiagnosticSource](https://www.nuget.org/packages/System.Diagnostics.DiagnosticSource/) but I was getting a conflict so I added it explicitly.

bruno-garcia replied:
Worth adding a comment above the package reference to point that out?

CodeBlanch replied:
Ended up removing the ref.

CodeBlanch commented on src/OpenTelemetry/Logs/LogRecord.cs:
```diff
@@ -79,23 +80,106 @@ public sealed class LogRecord
+            if (this.bufferedScopes != null)
+            {
+                ScopeForEachState<TState> forEachState = new ScopeForEachState<TState>(callback, state, 0);
+                IntPtr statePointer = new IntPtr(Unsafe.AsPointer(ref forEachState));
```
Why use unsafe?

The .NET API is:

```csharp
void ForEachScope<TState>(Action<object, TState> callback, TState state)
```

We would be in better shape if it was:
```csharp
void ForEachScope<TState>(Action<object, TState> callback, ref TState state)
   where TState : struct
```

As it is, you can't pass a mutable struct, which is needed to track depth. What the unsafe does is build that struct on the stack. Then pass a pointer into the method to the stack. Then build a ref to that pointer. Basically doing the `ref TState state`. This is unsafe because if someone took that pointer and tried to access the stack later, it would be bad. But we're not doing that, everything executes inline here.

reyang replied:
I'm trying to confirm if `forEachState` will be allocated on stack and not moved by GC since I couldn't find any document which provides the guarantee.

If I use GCHandle, there will be exception saying `Object contains non-primitive or non-blittable data`.

reyang replied:
For example, the following code will generate compile error:

```csharp
ScopeForEachState<TState>* p = stackalloc ScopeForEachState<TState>[1];
```

```error CS0208: Cannot take the address of, get the size of, or declare a pointer to a managed type ('LogRecord.ScopeForEachState<TState>')```

CodeBlanch replied:
@reyang I would be lying if I told you my understanding of this is 100% rock-solid 😄 If you can tap someone from runtime team to check this work that would be very helpful.

My _understanding_ is that because `ScopeForEachState` is a value type it will be on the stack. No GC, no moving. It won't go onto the heap unless we box it for some reason, which we aren't doing. So it is on the stack, then we call another method, passing in that stack pointer. In the new method it will still be at the same point on the stack so the pointer should be good.

The error on...

```ScopeForEachState<TState>* p = stackalloc ScopeForEachState<TState>[1];```

...I think is more because the struct contains a delegate & a generic. The compiler doesn't know how much space it needs to allocate for that.

If you have something more primitive like...

```csharp
public struct FixedLengthStruct
{
   public int Field1;
   public long Field2;
}
```

Then it _should_ work fine:

```FixedLengthStruct* p = stackalloc FixedLengthStruct[1];```

reyang replied:
Your explanation aligns with _my understanding_ (but I'm not 100% if my understanding is solid) 😄

noahfalk replied:
> This is unsafe because if someone took that pointer and tried to access the stack later, it would be bad. But we're not doing that, everything executes inline here.

That is one reason it is unsafe, but unfortunately not the only one. Another reason it is unsafe is because it prevents the GC from accurately tracking what objects are allowed to be collected. For example:

````C#
struct State
{
    object o;
}

void Run() 
{
    State s = new State();
    s.o = new object();
    IntPtr p = Unsafe.AsPointer(ref s);
    
    GC.Collect(); // the object in s.o may get collected because at this point in the method there are no
                  // references to it. It's easy to imagine that the local stack variable s should keep it alive
                  // but the JIT can see that there are no reads from s for the remainder of this method
                  // and if the code can't observe it then it doesn't need to be there. Whether or not it is
                  // depends on precise implementation details of the JIT and GC.

    ref State r = ref Unsafe.AsRef<State>(p);
    Console.WriteLine(r.o.ToString());              // r.o could be null or it could be a dangling pointer
                                                    // calling ToString() on it won't be good.
}
````
The code here is doing this pattern, where the object is 'callback'.


If you want state to give you mutable storage but don't want to allocate each time I'd suggest lazily allocating a singleton thread-local state object and then pass that as your TState.

CodeBlanch replied:
@noahfalk Great insight thanks for the help on this!

One tweak to your snippet:

```csharp
struct State
{
    object o;
}

void Run(object callback) 
{
    State s = new State();
    s.o = callback;
    IntPtr p = Unsafe.AsPointer(ref s);
    
    GC.Collect(); 

    ref State r = ref Unsafe.AsRef<State>(p);
    Console.WriteLine(r.o.ToString());
}
```

Basically, callback is passed into the method, not newed inside it. Does that change anything as far as the GC tracking is concerned?

I did consider the thread-local but figured it would add more overhead than avoiding the allocation would be worth. I'll do some benchmarking with it.

noahfalk replied:
Glad to help : )

> Basically, callback is passed into the method, not newed inside it. Does that change anything as far as the GC tracking is concerned?

The JIT is free to shorten the lifetime of a parameter the same way it shortens the lifetime of a local variable unfortunately. If the user calls it like this you could still be in trouble:
````C#
Run(new Callback());
````

CodeBlanch replied:
We discussed this on our weekly call today and decided the depth tracking wasn't a compelling enough feature to justify the complexity so I just removed it. Thanks for killing my puppy @noahfalk 🤣 J/k we can add it back on `LogRecordScope` in the future if we decide it makes sense to do so.

reyang commented on docs/logs/extending-the-sdk/MyExporter.cs:
```diff
@@ -46,8 +46,15 @@ public override ExportResult Export(in Batch<LogRecord> batch)
+            {
+                if (scope.Depth > 0)
+                {
+                    builder.Append(',');
```
```suggestion
                    builder.Append(", ");
```

CodeBlanch replied:
Updated

reyang commented on src/OpenTelemetry.Exporter.Console/ConsoleLogRecordExporter.cs:
```diff
@@ -64,10 +65,20 @@ public override ExportResult Export(in Batch<LogRecord> batch)
 
-                static void ProcessScope(object scope, ConsoleLogRecordExporter exporter)
-                    => exporter.WriteLine($"{"LogRecord.Scope:".PadRight(RightPaddingLength)}{scope}");
+                static void ProcessScopeItem(LogRecordScope scope, ConsoleLogRecordExporter exporter)
```
Curious: which one is more "native" to .NET - `ProcessScopeItem` or `ProcessScopeItems` (plural form)?

CodeBlanch replied:
I think the singular here is fine because it is getting called once for each scope and not being passed a list/enumerable of scopes? I did rename it back to "ProcessScope" (instead of "ProcessScopeItem"). That was a carry-over from me messing with the API.

Let me know if you want it plural, I don't feel strongly about it.

reyang replied:
I don't have strong preference here, it was just coming from curiosity.

CodeBlanch commented:
Updates based on SIG discussion today:

* BufferLogScopes is now internal.
* Removed depth tracking from LogRecordScope.
* Made LogRecordScope readonly struct instead of readonly ref struct and removed the delegate.

reyang reviewed (approved):
LGTM.

cijothomas reviewed (approved):
Thanks!
